txpipe/source_selector.py

```diff
@@ -673,9 +469,252 @@ def select_2d(self, data, calling_from_select=False):
             print("total 2D", sel.sum())
         return sel
 
+class TXSourceSelectorMetacal(TXSourceSelectorBase):
+    name='TXSourceSelectorMetacal'
+    def data_iterator(self):
+        bands = self.config['bands']
+        shear_cols = metacal_variants('mcal_T', 'mcal_s2n', 'mcal_g1', 'mcal_g2', 'mcal_flags')
+        shear_cols += ['ra', 'dec', 'mcal_psf_T_mean', 'weight']
+        shear_cols += band_variants(bands, 'mcal_mag', 'mcal_mag_err', shear_catalog_type='metacal')
+
+        if self.config['input_pz']:
+            shear_cols += metacal_variants('mean_z')
+        elif self.config['true_z']:
+            shear_cols += ['redshift_true']
+
+        chunk_rows = self.config['chunk_rows']
+        return self.iterate_hdf('shear_catalog', 'shear', shear_cols, chunk_rows)
+
+    def setup_output(self):
+        outfile = super().setup_output()
+        n = outfile['tomography/source_bin'].size
+        nbin_source = outfile['tomography/source_counts'].size
+        group = outfile.create_group('response')
+        group.create_dataset('R_gamma', (n,2,2), dtype='f')
+        group.create_dataset('R_S', (nbin_source,2,2), dtype='f')
+        group.create_dataset('R_gamma_mean', (nbin_source,2,2), dtype='f')
+        group.create_dataset('R_total', (nbin_source,2,2), dtype='f')
+        group.create_dataset('R_S_2d', (2,2), dtype='f')
+        group.create_dataset('R_gamma_mean_2d', (2,2), dtype='f')
+        group.create_dataset('R_total_2d', (2,2), dtype='f')
+        return outfile
+
+    def setup_response_calculators(self, nbin_source):
+        delta_gamma = self.config['delta_gamma']
+        calculators = [MetacalCalculator(self.select, delta_gamma) for i in range(nbin_source)]
+        calculators.append(MetacalCalculator(self.select_2d, delta_gamma))
+        return calculators
+
+    def write_tomography(self, outfile, start, end, source_bin, R):
+        super().write_tomography(outfile, start, end, source_bin, R)
+        group = outfile['response']
+        group['R_gamma'][start:end,:,:] = R
+
+    def compute_per_object_response(self, data):
+        delta_gamma = self.config['delta_gamma']
+        n = data['mcal_g1_1p'].size
+        R = np.zeros((n, 2, 2))
+        R[:,0,0] = (data['mcal_g1_1p'] - data['mcal_g1_1m']) / delta_gamma
+        R[:,0,1] = (data['mcal_g1_2p'] - data['mcal_g1_2m']) / delta_gamma
+        R[:,1,0] = (data['mcal_g2_1p'] - data['mcal_g2_1m']) / delta_gamma
+        R[:,1,1] = (data['mcal_g2_2p'] - data['mcal_g2_2m']) / delta_gamma
+        return R
+
+    def apply_simple_redshift_cut(self, data):
+        # If we have the truth pz then we just need to do the binning once,
+        # as in the parent class
+        if not self.config['input_pz']:
```

Comment:
do you mean true_z here instead of input_pz?

Reply:
It should be `not input_z` which here is the same as `true_z`. I've tidied up and explained.